components/spi_flash/test/test_read_write.c

```diff
@@ -0,0 +1,207 @@
+// Copyright 2010-2016 Espressif Systems (Shanghai) PTE LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Test for spi_flash_{read,write}.
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/param.h>
+
+#include <unity.h>
+#include <esp_spi_flash.h>
+#include <rom/spi_flash.h>
+#include "../cache_utils.h"
+#include "soc/timer_group_struct.h"
+#include "soc/timer_group_reg.h"
+
+/* Base offset in flash for tests. */
+#define FLASH_BASE 0x100000
+
+static void fill(char *dest, int32_t start, int32_t len)
+{
+    for (int32_t i = 0; i < len; i++)
+    {
+        *(dest + i) = (char) (start + i);
+    }
+}
+
+static int cmp_or_dump(const void *a, const void *b, size_t len)
+{
+    int r = memcmp(a, b, len);
+    if (r != 0)
+    {
+        for (int i = 0; i < len; i++)
+        {
+            fprintf(stderr, "%02x", ((unsigned char *) a)[i]);
+        }
+        fprintf(stderr, "\n");
+        for (int i = 0; i < len; i++)
+        {
+            fprintf(stderr, "%02x", ((unsigned char *) b)[i]);
+        }
+        fprintf(stderr, "\n");
+    }
+    return r;
+}
+
+static void IRAM_ATTR test_read(int src_off, int dst_off, int len)
+{
+    uint32_t src_buf[16];
+    char dst_buf[64], dst_gold[64];
+    fprintf(stderr, "src=%d dst=%d len=%d\n", src_off, dst_off, len);
+    memset(src_buf, 0xAA, sizeof(src_buf));
+    fill(((char *) src_buf) + src_off, src_off, len);
+    ESP_ERROR_CHECK(spi_flash_erase_sector((FLASH_BASE + src_off) / SPI_FLASH_SEC_SIZE));
+    spi_flash_disable_interrupts_caches_and_other_cpu();
+    assert(SPIWrite(FLASH_BASE, src_buf, sizeof(src_buf)) == SPI_FLASH_RESULT_OK);
+    spi_flash_enable_interrupts_caches_and_other_cpu();
+    memset(dst_buf, 0x55, sizeof(dst_buf));
+    memset(dst_gold, 0x55, sizeof(dst_gold));
+    fill(dst_gold + dst_off, src_off, len);
+
+    ESP_ERROR_CHECK(spi_flash_read(FLASH_BASE + src_off, dst_buf + dst_off, len));
+    assert(cmp_or_dump(dst_buf, dst_gold, sizeof(dst_buf)) == 0);
+}
+
+TEST_CASE("Test spi_flash_read", "[spi_flash_read]")
+{
+#if 1
```

Comment:
done. however, i couldn't identify a suitable config category to add this to (there isn't currently one for tests or spi_flash), so i'll leave it to you.